pkg/agent/agentclient/stream.go

```diff
@@ -0,0 +1,203 @@
+package agentclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"k8s.io/klog"
+	"sigs.k8s.io/apiserver-network-proxy/proto/agent"
+	// "k8s.io/client-go/util/retry"
+)
+
+const (
+	defaultRetry    = 20
+	defaultInterval = 5 * time.Second
+)
+
+type ReconnectError struct {
+	internalErr error
+	errChan     <-chan error
+}
+
+func (e *ReconnectError) Error() string {
+	return "transient error: " + e.internalErr.Error()
+}
+
+func (e *ReconnectError) Wait() error {
+	return <-e.errChan
+}
+
+type RedialableAgentClient struct {
+	stream agent.AgentService_ConnectClient
+
+	// connect opts
+	address       string
+	opts          []grpc.DialOption
+	conn          *grpc.ClientConn
+	stopCh        chan struct{}
+	reconnTrigger chan error
+
+	// locks
+	sendLock   sync.Mutex
+	recvLock   sync.Mutex
+	reconnLock sync.Mutex
+
+	// Retry times to reconnect to proxy server
+	Retry int
+
+	// Interval between every reconnect
+	Interval time.Duration
+}
+
+func NewRedialableAgentClient(address string, opts ...grpc.DialOption) *RedialableAgentClient {
+	c := &RedialableAgentClient{
+		address:  address,
+		opts:     opts,
+		Retry:    defaultRetry,
+		Interval: defaultInterval,
+		stopCh:   make(chan struct{}),
+	}
+
+	_ = <-c.triggerReconnect()
+
+	go c.probe()
+
+	return c
+}
+
+func (c *RedialableAgentClient) probe() {
+	for {
+		select {
+		case <-c.stopCh:
+			return
+		case <-time.After(c.Interval):
+			// health check
+			if c.conn != nil && c.conn.GetState() == connectivity.Ready {
+				continue
+			} else {
+				klog.Infof("Connection state %v", c.conn.GetState())
+			}
+		}
+
+		klog.Info("probe failure: reconnect")
+		if err := <-c.triggerReconnect(); err != nil {
+			klog.Infof("probe reconnect failed: %v", err)
+		}
+	}
+}
+
+func (c *RedialableAgentClient) Send(pkt *agent.Packet) error {
+	c.sendLock.Lock()
+	defer c.sendLock.Unlock()
+
+	if err := c.stream.Send(pkt); err != nil {
+		if err == io.EOF {
+			return err
+		}
+		return &ReconnectError{
+			internalErr: err,
+			errChan:     c.triggerReconnect(),
+		}
+	}
+
+	return nil
+}
+
+func (c *RedialableAgentClient) RetrySend(pkt *agent.Packet) error {
+	err := c.Send(pkt)
+	if err == nil {
+		return nil
+	} else if err == io.EOF {
+		return err
+	}
+
+	if err2, ok := err.(*ReconnectError); ok {
+		err = err2.Wait()
+	}
+	if err != nil {
+		return err
+	}
+	return c.RetrySend(pkt)
+}
+
+func (c *RedialableAgentClient) triggerReconnect() <-chan error {
+	c.reconnLock.Lock()
+	defer c.reconnLock.Unlock()
+
+	if c.reconnTrigger != nil {
+		return c.reconnTrigger
+	}
+
+	c.reconnTrigger = make(chan error)
+	go c.reconnect()
+
+	return c.reconnTrigger
+}
+
+func (c *RedialableAgentClient) doneReconnect(err error) {
+	c.reconnLock.Lock()
+	defer c.reconnLock.Unlock()
+
+	c.reconnTrigger <- err
```

Comment:
I have some concerns about what happens when we have multiple clients with the channel.
Please see https://play.golang.org/p/iKP5oETRahJ.
As far as I can tell we ONLY awake the first listener on a given channel instance.

Reply:
seems like a bug. the the chan needs to be closed.

Reply:
the error msg chan needs to go thru a 1:n fanout basically. good catch.